appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm

```diff
@@ -800,25 +800,21 @@
 
 ;; Also unhygienic (see comment above about foreach)
 
+;; The structure of this macro is important. If the argument to
+;; call-with-current-continuation is a lambda expression, then Kawa
+;; attempts to optimize it. This optimization fails spectacularly when
+;; the lambda expression is tail-recursive (like ours is). By binding
+;; the lambda expression to a variable and then calling via the
+;; variable, the optimizer is not invoked and the code produced, while
+;; not optmized, is correct.
+
 (define-macro (while condition body . rest)
-  `(call-with-current-continuation
-    (lambda (*yail-break*)
-      (let *yail-loop* ()
-        (if ,condition
-            (begin
-              ;; Kawa attempts to be smart by detecting unreachable
-              ;; code and reporting it as an error. This causes an
-              ;; issue here if *yail-break* is determined to be always
-              ;; called within body-form, since *yail-loop* will
-              ;; unreachable. The mixing of `or` and `and` operators
-              ;; below fools the Kawa optimizer so that even in
-              ;; trivial cases it won't flag the call to *yail-loop*
-              ;; as unreachable. It also works around a bug in the
-              ;; tail-recursive optimizer that breaks when the while
-              ;; loop is empty.
-              (or (and ((lambda () (begin ,body . ,rest))) #f)
-                  (*yail-loop*)))
-            #!null)))))
+  `(let ((cont (lambda (*yail-break*)
+                 (let *yail-loop* ()
+                   (if ,condition
+                       (begin (begin ,body . ,rest)
+                              (*yail-loop*)))))))
```

Comment:
Minor nit: If we modify this so that the `let` returns `#!null` like in the previous version, then the `testYailWhile` does not need to be modified. I don't think we rely on the previous semantics anywhere, so we could also go with this version.